include/wil/registry_helpers.h

```diff
@@ -0,0 +1,979 @@
+                        if (FAILED(hr))
+                        {
+                            return get_value_with_type_policy::HResult(hr);
+                        }
+                    }
+
+                    DWORD bytes_allocated{ reg_value_type_info::get_buffer_size(return_value) };
+                    for (;;)
+                    {
+                        constexpr DWORD* null_type{ nullptr };
+                        DWORD data_size_bytes{ bytes_allocated };
+                        hr = HRESULT_FROM_WIN32(::RegGetValueW(
+                            m_key,
+                            subkey,
+                            value_name,
+                            get_value_flags_from_value_type(type),
+                            null_type,
+                            reg_value_type_info::get_buffer(return_value),
+                            &data_size_bytes));
+
+                        // some return types we can grow as needed - e.g. when writing to a std::wstring
+                        // will only compile the below for those types that support dynamically growing the buffer
+                        if
+#if defined(__cpp_if_constexpr)
+                            constexpr
+#endif
+                            (reg_value_type_info::supports_resize_buffer<R>())
+                        {
+                            // GetRegValueW will indicate the caller allocate the returned number of bytes in one of two cases:
+                            // 1. returns ERROR_MORE_DATA
+                            // 2. returns ERROR_SUCCESS when we gave it a nullptr for the out buffer
+                            const bool shouldReallocate = (hr == HRESULT_FROM_WIN32(ERROR_MORE_DATA)) || (SUCCEEDED(hr) && (reg_value_type_info::get_buffer(return_value) == nullptr) && (data_size_bytes > 0));
+                            if (shouldReallocate)
+                            {
+                                // verify if resize_buffer succeeded allocation
+                                hr = reg_value_type_info::resize_buffer(return_value, data_size_bytes);
+                                if (SUCCEEDED(hr))
+                                {
+                                    // if it succeeds, continue the for loop to try again
+                                    bytes_allocated = data_size_bytes;
+                                    continue;
+                                }
+                            }
+
+                            if (SUCCEEDED(hr))
+                            {
+                                const auto current_byte_size = reg_value_type_info::get_buffer_size(return_value);
+                                if (current_byte_size != data_size_bytes)
+                                {
+                                    hr = reg_value_type_info::resize_buffer(return_value, data_size_bytes);
+                                }
+                            }
+
+                            if (SUCCEEDED(hr))

```

Comment:
I advise against defensive `if(SUCCEEDED` checks.

1. Your *intent* is that the function will basically stop doing further work, which is more clearly expressed with an early return/early break.
2. I have to understand how `hr` is manipulated across this entire function before I can understand how this code will run.
3. It's verbose; a lot of space of this function is dedicated to checking HRESULTs, when early returns would achieve the same result.

Returning early & using const named HRESULTs should make the code a bit clearer.

Reply:
ok. there were multiple points of feedback here. so I reworked it with named HRs.